open_schools_platform/organization_management/teachers/services.py

```diff
@@ -0,0 +1,84 @@
+import uuid
+
+from phonenumber_field.phonenumber import PhoneNumber
+from rest_framework.exceptions import NotAcceptable, MethodNotAllowed
+
+from open_schools_platform.common.services import model_update, BaseQueryHandler
+from open_schools_platform.common.utils import filter_dict_from_none_values
+from open_schools_platform.organization_management.circles.models import Circle
+from open_schools_platform.photo_management.photos.services import create_photo
+from open_schools_platform.query_management.queries.models import Query
+from open_schools_platform.query_management.queries.services import query_update
+from open_schools_platform.organization_management.teachers.models import TeacherProfile, Teacher
+from open_schools_platform.user_management.users.models import User
+
+
+def create_teacher_profile(name: str, age: int = None, user: User = None,
+                           phone: PhoneNumber = None, photo: uuid.UUID = None) -> TeacherProfile:
+    if not photo:
+        photo = create_photo()
+    teacher_profile = TeacherProfile.objects.create_teacher_profile(
+        name=name,
+        age=age,
+        phone=phone,
+        photo=photo,
+        user=user
+    )
+    return teacher_profile
+
+
+def create_teacher(name: str, circle: Circle = None, teacher_profile: TeacherProfile = None) -> Teacher:
+    teacher = Teacher.objects.create_teacher(
+        name=name,
+        circle=circle,
+        teacher_profile=teacher_profile
+    )
+    return teacher
+
+
+def update_teacher_profile(*, teacher_profile: TeacherProfile, data) -> TeacherProfile:
+    non_side_effect_fields = ['age', 'name', 'phone', 'photo']
+    filtered_data = filter_dict_from_none_values(data)
+    teacher_profile, has_updated = model_update(
+        instance=teacher_profile,
+        fields=non_side_effect_fields,
+        data=filtered_data
+    )
+    return teacher_profile
+
+
+class TeacherProfileQueryHandler(BaseQueryHandler):
+    allowed_statuses = [Query.Status.ACCEPTED, Query.Status.DECLINED, Query.Status.SENT, Query.Status.IN_PROGRESS,
+                        Query.Status.CANCELED]
+
+    @staticmethod
+    def query_handler(query: Query, new_status: str, user: User):
+        BaseQueryHandler.query_handler_checks(TeacherProfileQueryHandler, query, new_status, user)
+
+        circle_access = user.has_perm("circles.circle_access", query.recipient)
```

Comment:
I have a question, am I right, that this query handler is for situations, where TeacherProfile sends query to circle, so circle can hire him as a teacher? Did product approve this scenario? If he did, then you can just resolve this conversation, we are going to need a second query handler for situations, where circle actually sends query to TeacherProfile but we will create it later.

Reply:
No it should be Cirlce send query to TeacherProfile. Product approve that/

Reply:
So product didn't approved that TeacherProfile can send query to Circle?
Because this query_handler can be used only when TeacherProfile sends query to Circle because it gets circle_access from query.recipient. If you wanted to make this query_handler for cases, where Circle sends query to TeacherProfile, then you need to get circle_access from query.sender and teacher_profile_access from query.recipient.

Correct me if I didn't get your idea.